File: src/main/java/ru/develgame/jsfgame/ChatBean.java
package ru.develgame.jsfgame;

import com.sun.xml.internal.ws.util.StringUtils;
import ru.develgame.jsfgame.domain.ChatMessage;
import ru.develgame.jsfgame.domain.Person;
import ru.develgame.jsfgame.jms.ChangesInformer;
import ru.develgame.jsfgame.jms.MessagesType;

import javax.ejb.*;
import javax.inject.Inject;
import java.util.ArrayList;
import java.util.List;

@Singleton
@ConcurrencyManagement(ConcurrencyManagementType.CONTAINER)
public class ChatBean {
    @Inject
    private ChangesInformer changesInformer;

    private final List<ChatMessage> messages = new ArrayList<>();

    private static final int MAX_MESSAGES_COUNT = 10;

    private static final int MAX_HEADER_SIZE = 15;

    @Lock(LockType.WRITE)
    public void addMessage(String header, String message) {
        if (header.length() > MAX_HEADER_SIZE) {
            header = header.substring(header.length() - MAX_HEADER_SIZE);
        }

        messages.add(new ChatMessage(header, message));

        if (messages.size() > MAX_MESSAGES_COUNT) {
            messages.remove(0);
        }

        changesInformer.sendMessage(MessagesType.CHAT);
    }

    @Lock(LockType.READ)
    public List<ChatMessage> getMessages() {
        return new ArrayList<>(messages);
    }

    public void personJoined(Person person) {
        addMessage(person.getName(), String.format("Person %s joined", person.getName()));
    }
}

File: src/main/kotlin/ru/develgame/jsfgame/ChatBean.kt
package ru.develgame.jsfgame

import ru.develgame.jsfgame.domain.ChatMessage
import ru.develgame.jsfgame.domain.Person
import ru.develgame.jsfgame.jms.ChangesInformer
import ru.develgame.jsfgame.jms.MessagesType
import javax.ejb.*
import javax.inject.Inject

@Singleton
@ConcurrencyManagement(ConcurrencyManagementType.CONTAINER)
open class ChatBean {
    @Inject
    private lateinit var changesInformer: ChangesInformer

    private val messages = mutableListOf<ChatMessage>()

    @Lock(LockType.WRITE)
    open fun addMessage(header: String, message: String) {
        if (header.length > MAX_HEADER_SIZE)
            header.substring(header.length - MAX_HEADER_SIZE)

        messages.add(ChatMessage(header, message))

        if (messages.size > MAX_MESSAGES_COUNT) {
            messages.removeAt(0)
        }

        changesInformer.sendMessage(MessagesType.CHAT)
    }

    @Lock(LockType.READ)
    open fun getMessages(): List<ChatMessage> = messages.toList()

    open fun personJoined(person: Person) {
        addMessage(person.name, "Person ${person.name} joined")
    }

    companion object {
        const val MAX_MESSAGES_COUNT = 10
        const val MAX_HEADER_SIZE = 15
    }
}